### DISCUSSION TITLE
QMC5883L 3-Axis Magnetic SensorsSupport

### DISCUSSION THREAD

**Discussion by Performance20** [arendst/Tasmota#16705] [Ideas]
Hi,

I want to observe my gasmeter at home. I have here in Germany a model (Piepenburg & others) which have a build magnet which gives the opportunity to detect the change of the counter.
As sensor I used the QMC5883L, because is cheap to get from china (3€) and works very well to dedect the magnet. I tried as well using hall sensors but there was to much noising which needs a really big effort in software.
I like  Tasmota very well and and wanted to use it together with the ESP also here, but Tasmota not support such sensors.
Therfore I checked to add the support, found another impementation and checked several ardruino libraries. From all together I wrote a new driver to support this sensor.
Sensormodul is now ready and installed on the gasmeter. I used a Wemos and but all together in self printed housing. 

Question from site is, make es sence to add the driver to the tasmota github repository?
I'm not so familar with Github.

Regards

Helge

![Bild 01 10 22 um 14 47](https://user-images.githubusercontent.com/20727016/193410480-11d67580-f47a-4b84-8810-26df8b1efad4.jpg)

**Comment by barbudor** [comment]
Unless I read too fast, there isn't yet any magnetometer in Tasmota so that could be a nice addition.
You need to publish your changes to a branch of Tasmota forked in your own account and from there submit a Pull Request so your code is reviewed before merge.
We may also kindly request some change if needed to fit with the "Tasmota standard" 🙄

**Reply by sfromis** [reply]
Nothing merged for:
https://github.com/arendst/Tasmota/pull/5165
https://github.com/arendst/Tasmota/pull/14714

**Comment by Performance20** [comment]
OK, I will try it. Always good to learn new things :-)
One additonal question to the process of languages translation. I modified as well the the Web output, means add new words to the #defined in the english and german language files. But to be honest I'm not able to do is with other languages, perhaps for Spanish but no more. What happens with the other languages?

Helge

**Reply by arendst** [reply]
Just add them in english for others to translate.

**Reply by sfromis** [reply]
Until translated, the other languages will see the english-language phrases

**Reply by Performance20** [reply]
To be honest; I added to the de_DE.h and en_GB.h the new words I needed, compiling works well but if I try to compile for exampel the CN language version I got an error because the #defines are not there.
Can somebody explain how can solve the problem?

Helge

**Reply by barbudor** [reply]
You have to add the #define (in english) to each and every language file
That may be a bit tedious but that's ow it works
Sorry

**Reply by Performance20** [reply]
Ok thanks, I am doing it right now, a really stupid work.

**Comment by Performance20** [comment]
Then one additional question, sorry for that. I used for the driver file (they are numbered) the 99 because this number was free, but now no longer now. There ist a free number 33. Can I used this number for the driver (and name) or is there are rule how these numbers are assigned?

Helge

**Reply by arendst** [reply]
Pls use 119

**Reply by arendst** [reply]
No better use 33. Sorry for the confusion.

**Comment by Performance20** [comment]
One additional question. Is there a easy way to build all enviroments (languages version, ESP32, ...) in one shot ,like it was possible with a classical makefile with command "build all" . I use VSC as editor with Platformio.

Helge

**Reply by barbudor** [reply]
Normally if you open the PlatformIO Menu from the left (the alien icon)
On top you would have "PROJECT TASKS"
Open it and double click on "Build all"
That should build each and every target environment that are enabled (not commented) in platformio.Ini
Which means every standard build of Tasmota that are run daily
![image](https://user-images.githubusercontent.com/2996042/193633787-3de98ca8-3a4f-4112-af57-c57a5c8c57c8.png)

**Comment by BoBeRzE** [comment]
Hi,

i also want to use this sensor. Is it already integrated in the DEV-Brunch so i can test it?

Thank you

**Reply by barbudor** [reply]
It was merged last week
https://github.com/arendst/Tasmota/pull/16714

**Comment by BoBeRzE** [comment]
I did a little test with my board (HMC5883L). It seems to work. But I miss direction. Are special settings required for this board

![image](https://user-images.githubusercontent.com/49961176/195380255-786e8195-45da-48bb-9964-3529a4029ae3.png)

Thank you for the work

**Reply by sfromis** [reply]
One [commit](https://github.com/arendst/Tasmota/commit/5998fe856a588712a46b99b9be0dd39e2c25f7cd) during the review process says "and reduce functionality to the most necessary", but I did not see the discarding of the direction discussed in the thread.

**Comment by barbudor** [comment]
My understanding is that QmC5883 may be actually differrent from HMC which no more manufactured
So beware of Chinese naming and cloning

If you know an actual code/driver that work with your device we can have a look, compare and amend if needed

**Comment by Performance20** [comment]
Yes, I removed the whole compass staff from the driver to keep the driver small and simple. The other point is to have exact values you need an so called angle of declination which has to be to added in the calculation which depends from you location on the earth.
I think, would be possible better to calculate the compass value (heading degrees) outside of the driver, but it needs the atan2() function.
A good introduction about the topic you can find here:

https://www.circuitbasics.com/how-to-setup-a-magnetometer-on-the-arduino/

**Reply by christian-ffm** [reply]
Thank you for your work @Performance20. 

> Yes, I removed the whole compass staff from the driver to keep the driver small and simple. The other point is to have exact values you need an so called angle of declination which has to be to added in the calculation which depends from you location on the earth. I think, would be possible better to calculate the compass value (heading degrees) outside of the driver, but it needs the atan2() function. A good introduction about the topic you can find here:

Unfortunately, I am not successful in calculating the compass value. I only get values between ~150° and ~230°. Unfortunately, I cannot identify the problem. 

I do without the diclinationangle for the time being. I use the following formula: 
```
float heading=atan2(QMC5883L->MY, QMC5883L->MX);

 // Correct for when signs are reversed.
  if(heading < 0)
    heading += 2*PI;
    
  // Check for wrap due to addition of declination.
  if(heading > 2*PI)
    heading -= 2*PI;
   
  // Convert radians to degrees for readability.
  QMC5883L->HG = heading * 180/M_PI; 
```

Can someone help me here? In other Arduino scripts I always find a calibration process for the min/max values. I also don't do any conversion to gauss. Is that also necessary here?

**Reply by christian-ffm** [reply]
I managed to solve it on my own by using some code from the first commit. For everybody else:


Change datafield
```
    // data field
struct QMC5883L_s {
  int16_t   MX, MY, MZ, HG;
  int16_t   xhigh = 0, yhigh = 0, xlow = 0, ylow = 0;
  float     temp;
  uint16_t  scalar;
} *QMC5883L = nullptr;
```

Add Calculating
```
int16_t x = QMC5883L->MX;
int16_t y = QMC5883L->MY;

  // calculate azimut, heading
  if(x < QMC5883L->xlow) QMC5883L->xlow = x;
  if(x > QMC5883L->xhigh) QMC5883L->xhigh = x;
  if(y < QMC5883L->ylow) QMC5883L->ylow = y;
  if(y > QMC5883L->yhigh) QMC5883L->yhigh = y;

  if( QMC5883L->xlow == QMC5883L->xhigh || QMC5883L->ylow == QMC5883L->yhigh ) return;

  x -= (QMC5883L->xhigh + QMC5883L->xlow) / 2;
  y -= (QMC5883L->yhigh + QMC5883L->ylow) / 2;

 /* Rescale the measurement to the range observed. */
  float fx = (float) x / (QMC5883L->xhigh - QMC5883L->xlow);
  float fy = (float) y / (QMC5883L->yhigh - QMC5883L->ylow);

  x = -atan2(fy, fx) * 180.0 / M_PI;
  x += 180;  // no negative numbers
  
  QMC5883L->HG = x;
```

Change Output

```
#ifdef USE_WEBSERVER
const char HTTP_SNS_QMC5883L[] PROGMEM =
    "{s}QMC5883L " D_MX "{m}%d " D_UNIT_MICROTESLA "{e}"             // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
    "{s}QMC5883L " D_MY "{m}%d " D_UNIT_MICROTESLA "{e}"             // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
    "{s}QMC5883L " D_MZ "{m}%d " D_UNIT_MICROTESLA "{e}"             // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
    "{s}QMC5883L " D_HG "{m}%d " D_UNIT_DEGREE "{e}"                            // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
    "{s}QMC5883L " D_MAGNETICFLD "{m}%d " D_UNIT_MICROTESLA "{e}";   // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
#endif

void QMC5883L_Show(uint8_t json) {
  if (json) {
    ResponseAppend_P(PSTR(",\"QMC5883L\":{\"" D_JSON_MX "\":%d,\"" D_JSON_MY "\":%d,\"" D_JSON_MZ "\":%d,\"" D_JSON_HG "\":%d,\"" D_JSON_MAGNETICFLD "\":%u,\"" D_JSON_TEMPERATURE "\":%*_f}"),
      QMC5883L->MX, QMC5883L->MY, QMC5883L->MZ, QMC5883L->HG, QMC5883L->scalar, Settings->flag2.temperature_resolution, &QMC5883L->temp);
#ifdef USE_WEBSERVER
  } else {
    WSContentSend_PD(HTTP_SNS_QMC5883L, QMC5883L->MX, QMC5883L->MY, QMC5883L->MZ, QMC5883L->HG, QMC5883L->scalar);
    WSContentSend_Temp("QMC5883L", QMC5883L->temp);
#endif
  }
}
```


Again: Thank you very much for implementing!

**Comment by Performance20** [comment]
Yes, your first calculation was wrong, but now you used the code I removed. 
Most of the calucation is only to avoid outliers during measuremts.

Helge

**Comment by sebaskl** [comment]
Hi,
might be a dumb question, but unfortunately I couldn't find anything online. Could you guys give me a hint how I can read QMC5883L data using the Tasmota scripting? I've enabled USE_SCRIPT_I2C and am also able to read values using the ir(XXX) command. However, this doesn't provide any reasonable values. Any idea what I'm missing?
Thanks a lot in advance!

**Reply by barbudor** [reply]
Why using Scripting instead of the native driver ?

Anyhow, without sharing your script, we can just guess you made some mistakes

**Reply by sebaskl** [reply]
Hi barbudor,
the main reason I used scripting is that I can quickly modify and test it without compiling and flashing Tasmota everytime I want to change something.
I'm just using a very basic script for now:
```
>D
MZ=0
found=0

>B
found=ia(0x0D)

>S
if found>0 {
MZ=ir(0x05)
}

>W
MZ{m}%MZ%
```

**Reply by sebaskl** [reply]
This is the result:
![Tasmota](https://github.com/arendst/Tasmota/assets/147978932/48765d62-c0bf-499b-89ff-61a70effcfd4)

**Reply by barbudor** [reply]
Bad
You can't have both the native driver and Scripting accessing to the I2C device

Either you use the native driver and you get the values read fomr the native driver into your script `>T` section

Either you disable the native driver and use I2C commands to read the device
Please format your script properly to make it readable using triple-back-ticks

**Reply by sebaskl** [reply]
Thanks @barbudor! The >T section solved my issue

**Comment by skipper2020** [comment]
> Hi,
> 
> I want to observe my gasmeter at home. I have here in Germany a model (Piepenburg & others) which have a build magnet which gives the opportunity to detect the change of the counter. As sensor I used the QMC5883L, because is cheap to get from china (3€) and works very well to dedect the magnet. I tried as well using hall sensors but there was to much noising which needs a really big effort in software. I like Tasmota very well and and wanted to use it together with the ESP also here, but Tasmota not support such sensors. Therfore I checked to add the support, found another impementation and checked several ardruino libraries. From all together I wrote a new driver to support this sensor. Sensormodul is now ready and installed on the gasmeter. I used a Wemos and but all together in self printed housing.
> 
> Question from site is, make es sence to add the driver to the tasmota github repository? I'm not so familar with Github.
> 
> Regards
> 
> Helge
> 
> ![Bild 01 10 22 um 14 47](https://user-images.githubusercontent.com/20727016/193410480-11d67580-f47a-4b84-8810-26df8b1efad4.jpg)

Hi,
I'm new to Tasmota and want to do the same with my ESP and an GY-271 QMC5883L and my gas meter.
I want to ask you @Performance20 how do you interpret the turn arrounds from the magnetic wheel of the gas meter? Do you count them with a script on the Tasmota or do you interpret them in you Home Automation System?
I mostly want to use a script or Rule to do that, but I'm not sure if it's the best idea.
My idea is to have always the complete value of the the gas meter on the Tasmota and get it frequently by mqtt.
I don't know how to do that best.
Can someone help.

Beste regards

**Reply by Performance20** [reply]
Hi, 
I use a rule to build a trigger with hysteresis. If it match, I send a MQTT message to my home automation system (in my case FHEM). Tasmota needs to be compiled to support IF statements in rules. 

ON System#Boot DO Backlog Var1 0 ENDON    
 
ON QMC5883L#MagneticInduction<4000 DO    
 IF (Var1==0)      Var1 1;       
                              Var2=Var2+1;        
                              Var3=Var3+1;     
                              Mem1=Mem1+1;     
                              Event SendCnts   
ENDIF  
ENDON    
 
ON QMC5883L#MagneticInduction>20000 DO   
   IF (Var1==1)      Var1 0   ENDIF 
ENDON   
 
ON event#SendCnts DO    
  publish val/Gasmeter_87DB0D/GAS {"Time":"%timestamp%","Consumption":"0.1","Unit":"cbm","Counter":"%Var2%","CounterAll":"%Var3%", "Factor":"10", "CounterEEPROM":"%Mem1%"} 
ENDON

Regards Helge
